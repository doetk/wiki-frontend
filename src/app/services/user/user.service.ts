import { Injectable } from '@angular/core';
import { Http, Headers, Response, RequestOptions } from '@angular/http';
import { Observable } from 'rxjs/Observable';
import 'rxjs/add/observable/of';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/catch';
import 'rxjs/add/observable/throw';

import * as constants from '../../app.constants';
import { User } from '../../models/user';
import { NOOP } from '@angular/core/src/view/util';

const loginOptions = new RequestOptions({
  withCredentials: true,
});

@Injectable()
export class UserService {
  constructor(private http: Http) {}

  getById(_id: number) {
    const url = constants.USERSURL + '/' + _id;
    return this.http
      .get(url, loginOptions)
      .map(response => response.json())
      .catch(e => {
        return Observable.throw(e.json);
      });
  }

  getLastLogout(_id: number) {
    const url = constants.USERSURL + '/last-logout/' + _id;
    return this.http
      .get(url, loginOptions)
      .map(response => response.json())
      .catch(e => {
        return Observable.throw(e.json);
      });
  }

  login(username: string, password: string) {
    const body = `{"uid":"${username}","psw":"${password}"}`;
    const headers = new Headers();
    headers.append('Content-Type', 'application/json');
    const options = new RequestOptions({
      headers: headers,
      withCredentials: true,
    });

    return this.http
      .post(constants.LOGINURL, body, options)
      .map(response => response.json());
  }

  logout() {
    return this.http
      .put(constants.LOGOUTURL, null, loginOptions)
      .map(response => response.json());
  }

  signup(user: any) {
    return this.http
      .post(constants.SIGNUPURL, user)
      .map(response => response.json());
  }

  getAuthor(_id: number) {
    const url = constants.USERSURL + '/author/' + _id;
    return this.http.get(url).map(response => response.json());
  }

  updateInfo(user) {
    return this.http
      .put(constants.USERSURL + '/update/info', user, loginOptions)
      .map(res => res.json())
      .catch(e => {
        return Observable.throw(JSON.parse(e._body));
      });
  }

  updatePassword(user) {
    return this.http
      .put(
        constants.USERSURL + '/update/password',
        { user: user },
        loginOptions,
      )
      .map(res => res.json())
      .catch(e => {
        return Observable.throw(e.json);
      });
  }
}
