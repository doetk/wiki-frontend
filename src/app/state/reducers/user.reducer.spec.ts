import { loginFailed } from './../actions/user.action';
import { userReducer } from './user.reducer';
import { updateUserDetails } from '../actions/user.action';
import * as userActions from '../actions/user.action';
import { IUser } from '../interfaces/user.interface';
import * as fromUserReducer from './user.reducer';

describe('REDUCERS::userReducer:', () => {
  describe('An undefined action', () => {
    it('should return the default/intitial store state', () => {
      const action = {} as any;
      const result = userReducer(undefined, action);

      expect(result).toEqual(fromUserReducer.storeInitialState);
    });
  });

  describe('REQ_LOGIN', () => {
    it('should make loginInProgress to true', () => {
      const user = { username: 'test', password: '5432' };
      const createAction = userActions.reqLogin(user.username, user.password);
      const result = userReducer(
        fromUserReducer.storeInitialState,
        createAction,
      );

      expect(result.loginInProgress).toEqual(true);
      expect(result.username).toEqual('test');
      expect(result.loggedIn).toEqual(false);
    });
  });

  describe('REQ_SIGNUP', () => {
    it('should make signUpProgress to true', () => {
      const user = {
        username: 'test',
        password: '123',
        first_name: 'theo',
        last_name: 'Sam',
        email: 'abc@def.gh',
      };
      const createAction = userActions.reqSignup(
        user.username,
        user.password,
        user.first_name,
        user.last_name,
        user.email,
      );

      const expectedResult: IUser = {
        userid: undefined,
        username: 'test',
        password: '123',
        first_name: 'theo',
        last_name: 'Sam',
        email: 'abc@def.gh',
        loggedIn: false,
        loginInProgress: false,
        logoutInProgress: false,
        lastLoginFailed: false,
        lastLogoutFailed: false,
        signupInProgress: true,
        signedUp: false,
        updateInfo: false,
      };

      const result = userReducer(
        fromUserReducer.storeInitialState,
        createAction,
      );

      expect(result).toEqual(expectedResult);
      expect(result.username).toEqual('test');
      // expect(result.loggedIn).toEqual(true);
    });
  });
  // describe('LOGIN_COMPLETED', () => {
  //   it('should make loginInProgress to true', () => {
  //     const action = userActions.loginCompleted();
  //     const result = userReducer(fromUserReducer.storeInitialState, action);
  //     const expectedResult = {
  //       loginInProgress: false,
  //       lastLoginFailed: false,
  //       loggedIn: true,
  //       password: undefined,
  //     };
  //   });
  // });
});
//   describe('LOGIN_FAILURE', () => {
//     it('should have an error and no pending state', () => {
//       const error = 'login failed';
//       const createAction = new LoginFailure(error);

//       const expectedResult = {
//         error: error,
//         pending: false,
//       };

//       const result = reducer(fromLoginPage.initialState, createAction);

//       expect(result).toMatchSnapshot();
//     });
//   });
