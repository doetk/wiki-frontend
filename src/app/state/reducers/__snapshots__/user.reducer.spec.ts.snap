// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userReducer: REQ_LOGIN should make loginInProgress to true 1`] = `
Object {
  "email": undefined,
  "first_name": undefined,
  "lastLoginFailed": false,
  "lastLogoutFailed": false,
  "last_name": undefined,
  "loggedIn": false,
  "loginInProgress": true,
  "logoutInProgress": false,
  "password": "5432",
  "signedUp": false,
  "signupInProgress": false,
  "userid": undefined,
  "username": "test",
}
`;

exports[`userReducer: REQ_SIGNUP should make signUpProgress to true 1`] = `
Object {
  "email": "abc@def.gh",
  "first_name": "theo",
  "lastLoginFailed": false,
  "lastLogoutFailed": false,
  "last_name": "Sam",
  "loggedIn": false,
  "loginInProgress": false,
  "logoutInProgress": false,
  "password": "123",
  "signedUp": false,
  "signupInProgress": true,
  "userid": undefined,
  "username": "test",
}
`;

exports[`userReducer: undefined action should return the default state 1`] = `
Object {
  "email": undefined,
  "first_name": undefined,
  "lastLoginFailed": false,
  "lastLogoutFailed": false,
  "last_name": undefined,
  "loggedIn": false,
  "loginInProgress": false,
  "logoutInProgress": false,
  "signedUp": false,
  "signupInProgress": false,
  "userid": undefined,
  "username": undefined,
}
`;
